grpc/creds/creds.go

```diff
@@ -51,25 +54,148 @@ func (tc *transportCredentials) ClientHandshake(ctx context.Context, addr string
 	}
 }
 
-// ServerHandshake performs the TLS handshake for a server <- client connection
-func (tc *transportCredentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
-	return nil, nil, fmt.Errorf("boulder/grpc/creds: Server-side handshakes are not implemented")
+// ServerHandshake is not implemented for a `clientTransportCredentials`, use
+// a `serverTransportCredentials` if you require `ServerHandshake`.
+func (tc *clientTransportCredentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	return nil, nil, fmt.Errorf(
+		"boulder/grpc/creds: Server-side handshakes are not implemented with " +
+			"clientTransportCredentials")
 }
 
 // Info returns information about the transport protocol used
-func (tc *transportCredentials) Info() credentials.ProtocolInfo {
+func (tc *clientTransportCredentials) Info() credentials.ProtocolInfo {
 	return credentials.ProtocolInfo{
 		SecurityProtocol: "tls",
 		SecurityVersion:  "1.2", // We *only* support TLS 1.2
 	}
 }
 
 // GetRequestMetadata returns nil, nil since TLS credentials do not have metadata.
-func (tc *transportCredentials) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+func (tc *clientTransportCredentials) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
 	return nil, nil
 }
 
 // RequireTransportSecurity always returns true because TLS is transport security
-func (tc *transportCredentials) RequireTransportSecurity() bool {
+func (tc *clientTransportCredentials) RequireTransportSecurity() bool {
+	return true
+}
+
+// serverTransportCredentials is a grpc/credentials.TransportCredentials which supports
+// filtering acceptable peer connections by a list of accepted client certificate SANs
+type serverTransportCredentials struct {
+	serverConfig *tls.Config
+	acceptedSANs map[string]struct{}
+}
+
+// NewServerTransport returns a new initialized grpc/credentials.TransportCredentials for server usage
+func NewServerTransport(serverConfig *tls.Config, acceptedSANs map[string]struct{}) credentials.TransportCredentials {
+	return &serverTransportCredentials{serverConfig, acceptedSANs}
+}
+
+// validateClient checks a peer's client certificate's SAN entries against
+// a list of accepted SANs. If the client certificate does not have a SAN on the
+// list it is rejected.
+//
+// Note 1: This function *only* verifies the SAN entries! Callers are expected to
```

Comment:
> It's a little weird to split the functions this way - why not just move conn.Handshake() into this function so it doesn't need this precondition? Of course, with that and the nil check missing, it might make more sense to just outright combine validateClient and ServerHandshake.

It's significantly harder to test if I have to arrange for a full client/server TLS handshake. The goal was to remove everything that wasn't the handshake to a separate function to make it easier to test the functionality we're adding vs the TLS handshake.